crates/chia-datalayer/src/lib.rs

```diff
@@ -1,42 +1,155 @@
     pub blob: Vec<u8>,
-    pub kv_to_index: HashMap<Key, Index>,
-    pub free_indexes: Vec<Index>,
-    pub last_allocated_index: Index,
 }
 
+// TODO: fill out related to the serializations
+const METADATA_SIZE: u32 = 2;
+const DATA_SIZE: u32 = 0;
+const SPACING: u32 = METADATA_SIZE + DATA_SIZE;
+
+impl MerkleBlob {
+    pub fn get_raw_node(&self, index: TreeIndex) -> RawMerkleNode {
+        // TODO: handle invalid indexes?
+        // TODO: handle overflows?
+        let metadata_start = index * SPACING;
+        let data_start = metadata_start + METADATA_SIZE;
+        let end = data_start + DATA_SIZE;
+
+        let metadata_blob = &self.blob[metadata_start as usize..data_start as usize];
+        let data_blob = &self.blob[data_start as usize..end as usize];
+        let metadata = NodeMetadata::load(metadata_blob);
+        RawMerkleNode::load(metadata, 0, data_blob)
+    }
+}
+
+pub enum RawMerkleNode {
+    Root {
+        left: TreeIndex,
+        right: TreeIndex,
+        hash: Hash,
+        // TODO: kinda feels questionable having it be aware of its own location
+        // TODO: just always at zero?
+        index: TreeIndex,
+    },
+    Internal {
+        parent: TreeIndex,
+        left: TreeIndex,
+        right: TreeIndex,
+        hash: Hash,
+        // TODO: kinda feels questionable having it be aware of its own location
+        index: TreeIndex,
+    },
+    Leaf {
+        parent: TreeIndex,
+        key_value: KVId,
+        hash: Hash,
+        // TODO: kinda feels questionable having it be aware of its own location
+        index: TreeIndex,
+    },
+}
+
+impl RawMerkleNode {
+    // TODO: how do i say what i'm passing in is length two if i [u8; 2] here
+    pub fn load(metadata: NodeMetadata, index: TreeIndex, blob: &[u8]) -> Self {
+        match metadata.node_type {
+            NodeType::Internal => RawMerkleNode::Internal {
+                // TODO: get these right
+                parent: TreeIndex::from_be_bytes(<[u8; 4]>::try_from(&blob[0..4]).unwrap()),
+                left: TreeIndex::from_be_bytes(<[u8; 4]>::try_from(&blob[4..8]).unwrap()),
+                right: TreeIndex::from_be_bytes(<[u8; 4]>::try_from(&blob[8..12]).unwrap()),
+                hash: <[u8; 32]>::try_from(&blob[12..46]).unwrap(),
+                index,
+            },
+            NodeType::Leaf => RawMerkleNode::Leaf {
+                // TODO: this try from really right?
+                parent: TreeIndex::from_be_bytes(<[u8; 4]>::try_from(&blob[0..4]).unwrap()),
+                key_value: KVId::try_from(&blob[4..36]).unwrap(),
+                hash: Hash::try_from(&blob[36..68]).unwrap(),
+                index,
+            },
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
 pub struct NodeMetadata {
     pub node_type: NodeType,
+    pub dirty: bool,
+}
+
+impl NodeMetadata {
+    // TODO: how do i say what i'm passing in is length two if i [u8; 2] here
+    pub fn load(blob: &[u8]) -> Self {
```

Comment:
you can just make the argument `[u8; 2]` and pass it by value rather than reference, it'd be an array instead of a slice

Reply:
changed.